src/signer/server.rs

```diff
@@ -63,42 +76,81 @@ impl Signer for Ed25519Signer {
 }
 
 #[tokio::main]
-async fn main() -> Result<(), Box<dyn std::error::Error>> {
+async fn main() -> remote_signer::Result<()> {
     SimpleLogger::from_env().init().unwrap();
     let config_arg = App::new("Remote Signer")
-        .arg(Arg::with_name("config")
-             .short("c")
-             .long("config")
-             .takes_value(true)
-             .value_name("FILE")
-             .default_value("signer_config.json")
-             .help("Dispatcher .json configuration file")
-        ).get_matches();
+        .arg(
+            Arg::with_name("config")
+                .short("c")
+                .long("config")
+                .takes_value(true)
+                .value_name("FILE")
+                .default_value("signer_config.json")
+                .help("Dispatcher .json configuration file"),
+        )
+        .get_matches();
 
     let conf_path = config_arg.value_of("config").unwrap();
     info!("Parsing configuration file `{}`.", conf_path);
     let (config, keypairs) = config::parse_signer(conf_path)?;
     debug!("Parsed configuration file: {:?}", config);
     let addr = config.bind_addr.parse()?;
-    let server_cert = tokio::fs::read(&config.tls.cert).await?;
-    let server_key = tokio::fs::read(&config.tls.key).await?;
-    let server_identity = Identity::from_pem(server_cert, server_key);
-    let client_ca_cert = tokio::fs::read(&config.tls.ca).await?;
-    let client_ca_cert = Certificate::from_pem(client_ca_cert);
-    let tls = ServerTlsConfig::new()
-        .identity(server_identity)
-        .client_ca_root(client_ca_cert);
-
-    let signer = Ed25519Signer { config, keypairs };
+
+    let signer = Ed25519Signer {
+        keypairs: Arc::new(Mutex::new(keypairs)),
+    };
     debug!("Initialized Signer server: {:?}", signer);
 
-    info!("Serving on {}...", addr);
+    let signal = reload_configs_upon_signal(conf_path, Arc::clone(&signer.keypairs));
 
-    Server::builder()
-        .tls_config(tls)?
+    let serv = Server::builder()
         .add_service(SignerServer::new(signer))
         .serve(addr)
-        .await?;
+        .map_err(|error| RemoteSignerError::from(error));
+
+    info!("Serving on {}...", addr);
+
+    let result = future::try_join(signal, serv).await;
```

Comment:
Don't match result, use `await?` and `Ok(())`

Reply:
oh yeah, forgot to change it after I was experimenting

Reply:
I can't use `await?`  because main wants to return a `Result`.
I can't use `await` because it returns `Result<<(),()>, RemoteSignerError>`

So I can either:
1. Change what `Main`  returns, but I rather it stays `Result` and rust forces it be of the structure `Result<(), Err>`...
2. Use `map_ok` to change `().()` -> `()`. But apparently I need to implement several traits for it to work...

So I choose (3), leaving things as they are :-)

Reply:
Ok so let's at least get rid of the ugly variable :P
```
match future::try_join(serv, signal).await {
        Ok(_) => Ok(()),
        Err(e) => Err(e),
}
```